src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenEndpoint.java

```diff
@@ -271,6 +303,19 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	@SuppressWarnings("unchecked")
+	private void exchange(Message<?> message) {
+		Object replyObject = sendAndReceive(message);
+		V reply = null;
+		try {
+			reply = (V) replyObject;
+		}
+		catch (ClassCastException e) {
+			logger.error("Unexpected reply type: " + replyObject.getClass().getName(), e);
+		}
+		this.replyTemplate.send(getMessageBuilderFactory().withPayload(reply).build());
```

Comment:
Hm. I see you use `messageKey` in the `KafkaProducerMessageHandler` for correlation. 
But I don't see that `key` re-mapping here from the request `message`...